PP-1059: Jobs in queues with resources_available limits (e.g., resvs)…

bhroam opened the pull request:
… may not run due to equiv classes

<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-1059](https://pbspro.atlassian.net/browse/PP-1059)**

#### Problem description
* When a queue has a resources_available limit, it can cause other similar jobs to not run.

#### Cause / Analysis
* When a job hits a queue's resources_available limit, the equivalence class it is in is marked can not run.  Jobs in other queues can be in that class.  If they are, they'll also not run even though their queue does not have a resources_available limit.  This is because the class is already marked can not run.

#### Solution description
* Jobs in different queues can be in the same equivalence class when the queue does not matter.  Jobs in queues with resources_available or jobs in reservation queues need to be in their own equivalence class.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [X] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [X] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [X] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [X] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [X] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

agrawalravi90 commented on src/scheduler/data_types.h:
```diff
@@ -372,6 +372,7 @@ struct queue_info
 	unsigned has_soft_limit:1;	/* queue has a soft user/grp limit set */
 	unsigned has_hard_limit:1;	/* queue has a hard user/grp limit set */
 	unsigned is_peer_queue:1;	/* queue is a peer queue */
+	unsigned has_res_av:1;		/* queue has resources_available limits */
```
I think we should add the word 'limit' to the identifier, maybe 'has_resav_limit' ?

minghui-liu replied:
+1 on changing the name to `has_res_av_limit`.

bhroam replied:
I changed the name.

agrawalravi90 commented on test/tests/functional/pbs_equiv_classes.py:
```diff
@@ -1840,3 +1840,44 @@ def test_held_jobs_equiv_class(self):
+        jid2 = self.server.submit(J)
+
+        J = Job(TEST_USER)
+        jid3 = self.server.submit(J)
```
I don't like relying on the fact that PBS/PTL will add a default select to the job and would suggest that you add the select line explicitly.

bhroam replied:
I meant to include it.  I just missed it, thanks for catching this.

bhroam commented:
I modified the test a little.  I'm not sure why it was passing before.  Job 2 runs and Job 3 doesn't.  I was checking for them in the other order.  I also removed the expect() for the comment because it can change due to a race condition.  The initial cycle it gets set to the comment about not enough queue resources.  Subsequent cycles the node is full so it changes to not enough free nodes.

agrawalravi90 commented:
Thanks for addressing the comments, could you please attach the latest testing logs to the ticket?

bhroam commented:
@agrawalravi90 I attached the test logs and squashed.  When running the entire equivalence class test suite, I found another bug.  I filed it as PP-1127.

agrawalravi90 commented:
Thanks Bhroam, I'll merge this in
